platform/kube/registry.go

```diff
@@ -0,0 +1,331 @@
+	multierror "github.com/hashicorp/go-multierror"
+
+	"istio.io/manager/model"
+
+	"k8s.io/client-go/1.5/kubernetes"
+	"k8s.io/client-go/1.5/pkg/api"
+	"k8s.io/client-go/1.5/pkg/api/errors"
+	"k8s.io/client-go/1.5/pkg/api/unversioned"
+	"k8s.io/client-go/1.5/pkg/api/v1"
+	"k8s.io/client-go/1.5/pkg/apis/extensions/v1beta1"
+	"k8s.io/client-go/1.5/pkg/runtime"
+	"k8s.io/client-go/1.5/pkg/runtime/serializer"
+	"k8s.io/client-go/1.5/rest"
+	"k8s.io/client-go/1.5/tools/clientcmd"
+)
+
+const IstioAPIGroup = "istio.io"
+const IstioResourceVersion = "v1"
+
+// KubernetesRegistry bindings for the manager:
+// - configuration objects are stored as third-party resources
+type KubernetesRegistry struct {
+	mapping   model.KindMap
+	client    *kubernetes.Clientset
+	dyn       *rest.RESTClient
+	Namespace string
+}
+
+// CreateRESTConfig for cluster API server, pass empty config file for in-cluster
+func CreateRESTConfig(kubeconfig string, km model.KindMap) (*rest.Config, error) {
+	config, err := clientcmd.BuildConfigFromFlags("", kubeconfig)
+	if err != nil {
+		return nil, err
+	}
+
+	version := unversioned.GroupVersion{
+		Group:   IstioAPIGroup,
+		Version: IstioResourceVersion,
+	}
+
+	config.GroupVersion = &version
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				version,
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			for kind := range km {
+				scheme.AddKnownTypeWithName(unversioned.GroupVersionKind{
+					Group:   IstioAPIGroup,
+					Version: IstioResourceVersion,
+					Kind:    kind,
+				}, &Config{})
+				scheme.AddKnownTypeWithName(unversioned.GroupVersionKind{
+					Group:   IstioAPIGroup,
+					Version: IstioResourceVersion,
+					Kind:    kind + "List",
+				}, &ConfigList{})
+			}
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	return config, nil
+}
+
+func NewKubernetesRegistry(kubeconfig string, km model.KindMap) (*KubernetesRegistry, error) {
+	config, err := CreateRESTConfig(kubeconfig, km)
+	if err != nil {
+		return nil, err
+	}
+	cl, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	dyn, err := rest.RESTClientFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	out := &KubernetesRegistry{
+		mapping:   km,
+		client:    cl,
+		dyn:       dyn,
+		Namespace: api.NamespaceDefault,
+	}
+
+	if err = out.RegisterResources(); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// RegisterResources creates third party resources
+func (kr *KubernetesRegistry) RegisterResources() error {
+	var out error
+	for kind, v := range kr.mapping {
+		apiName := KindToAPIName(kind)
+		// initialize resource if it does not exist
+		res, err := kr.client.Extensions().ThirdPartyResources().Get(apiName)
+		if err != nil {
```

Comment:
commented on this a bunch on the previous commit, but we should really strive to reduce the nesting here.

something like:

```
if err == nil {
    log...
    continue
}
if !errors.IsNotFound(err) {
    out = ...
    continue
}
tpr := &....
...
```